Sources/SecureEnclaveValet.swift

```diff
@@ -0,0 +1,286 @@
+    /// - returns: `false` if the keychain is not accessible.
+    @objc
+    @discardableResult
+    public func removeAllObjects() -> Bool {
+        return execute(in: lock) {
+            switch Keychain.removeAllObjects(matching: keychainQuery) {
+            case .success:
+                return true
+                
+            case .error:
+                return false
+            }
+        }
+    }
+    
+    /// Migrates objects matching the input query into the receiving SecureEnclaveValet instance.
+    /// - parameter query: The query with which to retrieve existing keychain data via a call to SecItemCopyMatching.
+    /// - parameter removeOnCompletion: If `true`, the migrated data will be removed from the keychain if the migration succeeds.
+    /// - returns: Whether the migration succeeded or failed.
+    /// - note: The keychain is not modified if a failure occurs.
+    @objc(migrateObjectsMatchingQuery:removeOnCompletion:)
+    public func migrateObjects(matching query: [String : AnyHashable], removeOnCompletion: Bool) -> MigrationResult {
+        return execute(in: lock) {
+            return Keychain.migrateObjects(matching: query, into: keychainQuery, removeOnCompletion: removeOnCompletion)
+        }
+    }
+    
+    /// Migrates objects matching the vended keychain query into the receiving SecureEnclaveValet instance.
+    /// - parameter keychain: An objects whose vended keychain query is used to retrieve existing keychain data via a call to SecItemCopyMatching.
+    /// - parameter removeOnCompletion: If `true`, the migrated data will be removed from the keychfain if the migration succeeds.
+    /// - returns: Whether the migration succeeded or failed.
+    /// - note: The keychain is not modified if a failure occurs.
+    @objc(migrateObjectsFromKeychain:removeOnCompletion:)
+    public func migrateObjects(from keychain: KeychainQueryConvertible, removeOnCompletion: Bool) -> MigrationResult {
+        return migrateObjects(matching: keychain.keychainQuery, removeOnCompletion: removeOnCompletion)
+    }
+    
+    // MARK: Private Properties
+    
+    private let service: Service
+    private let lock = NSLock()
+    private let keychainQuery: [String : AnyHashable]
+}
+
+
+// MARK: - Objective-C Compatibility
+
+
+extension SecureEnclaveValet {
+    
+    // MARK: Public Class Methods
+    
+    /// - parameter identifier: A non-empty string that uniquely identifies a SecureEnclaveValet.
+    /// - returns: A SecureEnclaveValet that reads/writes keychain elements with the desired flavor.
+    @objc(valetWithIdentifier:accessControl:)
+    public class func 🚫swift_valet(with identifier: String, accessControl: SecureEnclaveAccessControl) -> SecureEnclaveValet? {
+        guard let identifier = Identifier(nonEmpty: identifier) else {
+            return nil
+        }
+        return valet(with: identifier, accessControl: accessControl)
```

Comment:
I think the guard/return is a bit more clear. Didn't realize that `map` could be used on non-collections to transform a single element (TIL!). Certainly a nifty suggestion though!

Reply:
Ah yeah `Optional.map` is super neat. It's not _exactly_ that you're working with single elements, it's actually available because `Optional` is a 👻 FUNCTOR 👻. The definition is basically:

```swift
extension Optional {
  func map<U>(_ transform: (Wrapped) -> U) -> U? {
    switch self {
      case let .some(value): return transform(value)
      case .none: return .none
    }
  }
}
```

(The [actual implementation](https://github.com/apple/swift/blob/master/stdlib/public/core/Optional.swift#L140-L171) handles throwing functions too).


Note that you _also_ [have access to `Optional.flatMap`](https://github.com/apple/swift/blob/master/stdlib/public/core/Optional.swift#L173-L202), which is equally cool and handy.

Reply:
That's freaking sweet. Thanks for the documentation!!

Is this a standard you've seen used elsewhere? Seems like a really nifty (if slightly unexpected) way to enhance control flow. I half way to steer clear of it because it's a bit unexpected... but the other half of me wants to use the new hotness

Reply:
Definitely. I'm a fan of it in general, but I'm a bit of a functional programming nerd. It's especially useful for deferring your need to care if something exists or not until the last possible moment. For example, you could imagine a (n incredibly convoluted) situation where you have an optional string and want to perform some (incredibly convoluted) transformations on it before assigning it to a non-optional property:

```swift
let optString: String? = // maybe I exist, maybe I don't

let transformedString = optString
  .map(uppercase) // uppercase: (String) -> String
  .map(lowercase) // lowercase: (String) -> String
  .map(titlecase) // titlecase: (String) -> String
  .flatMap(ensureNonEmpty) // ensureNonEmpty: (String) -> String?

let string: String = transformedString ?? "default"
```

Reply:
SHAMELESS SELF PROMOTION: I did [a talk about functors at NSNorth 2015](https://github.com/gfontenot/talks/tree/master/Functors) if you're interested in more